rspace/src/main/scala/coop/rchain/rspace/history/LMDBTrieStore.scala

```diff
@@ -92,15 +111,57 @@ class LMDBTrieStore[K, V] private (val env: Env[ByteBuffer],
       throw new Exception(s"could not persist: $hash")
     }
   }
+
+  private[this] def getAllCheckpoints(txn: Txn[ByteBuffer]): Seq[Blake2b256Hash] =
+    withResource(_dbCheckpoints.iterate(txn)) { (it: CursorIterator[ByteBuffer]) =>
+      it.asScala.foldLeft(Seq.empty[Blake2b256Hash]) { (acc, keyVal) =>
+        acc ++ Codec[Seq[Blake2b256Hash]].decode(BitVector(keyVal.`val`())).map(_.value).get
+      }
+    }
+
+  private[this] def getCheckpointsInBranch(txn: Txn[ByteBuffer],
+                                           branch: Branch): Seq[Blake2b256Hash] = {
+    val encodedBranch     = Codec[Branch].encode(branch).get
+    val encodedBranchBuff = encodedBranch.bytes.toDirectByteBuffer
+    Option(_dbCheckpoints.get(txn, encodedBranchBuff))
+      .map { bytes =>
+        Codec[Seq[Blake2b256Hash]].decode(BitVector(bytes)).map(_.value).get
+      }
+      .getOrElse(Seq.empty)
+  }
+
+  private[rspace] def validateCheckpointAndPutRoot(txn: Txn[ByteBuffer],
+                                                   branch: Branch,
+                                                   hash: Blake2b256Hash): Unit =
+    getRoot(txn, branch)
+      .find(_ == hash)
+      .orElse {
+        getCheckpointsInBranch(txn, branch)
+          .find(_ == hash)
+          .map { blake: Blake2b256Hash =>
+            putRoot(txn, branch, blake)
+            blake
+          }
+      }
+      .orElse {
+        getAllCheckpoints(txn)
+          .find(_ == hash)
+          .map { blake: Blake2b256Hash =>
+            putRoot(txn, branch, blake)
+            blake
+          }
+      }
+      .getOrElse(throw new Exception(s"Unknown checkpoint."))
 }
 
 object LMDBTrieStore {
 
   def create[K, V](env: Env[ByteBuffer])(implicit
                                          codecK: Codec[K],
                                          codecV: Codec[V]): LMDBTrieStore[K, V] = {
-    val dbTrie: Dbi[ByteBuffer]  = env.openDbi("Trie", MDB_CREATE)
-    val dbRoots: Dbi[ByteBuffer] = env.openDbi("Roots", MDB_CREATE)
-    new LMDBTrieStore[K, V](env, dbTrie, dbRoots)
+    val dbTrie: Dbi[ByteBuffer]        = env.openDbi("Trie", MDB_CREATE)
+    val dbRoots: Dbi[ByteBuffer]       = env.openDbi("Roots", MDB_CREATE)
+    val dbCheckpoints: Dbi[ByteBuffer] = env.openDbi("Checkpoints", MDB_CREATE)
```

Comment:
Some reservations about using the term "checkpoint" here as we have the related type `Checkpoint`.

Don't have a specific replacement in mind but thinking about things like "branchHistories", "branchLogs", etc.